File: hyze_threat_scanner.cpp
// Hyze AI Threat Scanner v1.0 - C++ Production
// Model/dataset/I/O scanning + real-time blocking

#include <filesystem>
#include <fstream>
#include <vector>
#include <string>
#include <nlohmann/json.hpp>
#include "hyze_ipu_client.h"

using json = nlohmann::json;

class HyzeThreatScanner {
private:
    HyzeIpuClient ipu;
    std::unordered_map<std::string, std::vector<uint8_t>> signatures;
    std::string quarantine_dir;

public:
    struct Threat {
        std::string severity;
        std::string type;
        std::string file_path;
        std::string details;
        float confidence;
    };

    HyzeThreatScanner(const std::string& quarantine = "./quarantine") 
        : quarantine_dir(quarantine), ipu("pci:10ee:7021") {
        std::filesystem::create_directory(quarantine_dir);
        load_signatures("threat_signatures.json");
    }

    std::vector<Threat> scan_model(const std::string& model_path) {
        std::vector<Threat> threats;
        auto model_bytes = read_file(model_path);

        // Signature matching
        for (const auto& [name, sig] : signatures) {
            if (find_pattern(model_bytes, sig)) {
                threats.push_back({
                    "CRITICAL", "Backdoor", model_path, 
                    "Signature: " + name, 1.0f
                });
            }
        }

        // IPU behavioral scan
        auto features = extract_model_features(model_bytes);
        float threat_score = ipu.model_threat_scan(features);
        
        if (threat_score > 0.9f) {
            threats.push_back({
                "HIGH", "Poisoning", model_path,
                "Gradient poisoning: " + std::to_string(threat_score), 
                threat_score
            });
        }

        return threats;
    }

    std::vector<Threat> scan_dataset(const std::string& dataset_dir) {
        std::vector<Threat> threats;
        
        for (const auto& entry : std::filesystem::recursive_directory_iterator(dataset_dir)) {
            if (entry.is_regular_file()) {
                auto file_bytes = read_file(entry.path().string());
                
                if (check_data_poisoning(file_bytes)) {
                    threats.push_back({
                        "HIGH", "Poisoning", entry.path().string(),
                        "Training data poisoning", 0.95f
                    });
                }
            }
        }
        
        return threats;
    }

    std::vector<Threat> scan_io(const std::vector<uint8_t>& input, 
                               const std::vector<uint8_t>& output) {
        std::vector<Threat> threats;

        if (prompt_injection_detected(input)) {
            threats.push_back({
                "CRITICAL", "Jailbreak", "input_stream",
                "Prompt injection attack", 1.0f
            });
        }

        if (pii_leakage_detected(output)) {
            threats.push_back({
                "HIGH", "DataLeak", "output_stream",
                "PII leakage detected", 0.98f
            });
        }

        return threats;
    }

    void block_threats(const std::vector<Threat>& threats) {
        for (const auto& threat : threats) {
            if (threat.type == "Backdoor") {
                quarantine_file(threat.file_path);
            } else if (threat.type == "Jailbreak") {
                ipu.block_stream();
            }
        }
    }

private:
    std::vector<uint8_t> read_file(const std::string& path) {
        std::ifstream file(path, std::ios::binary | std::ios::ate);
        auto size = file.tellg();
        file.seekg(0);
        
        std::vector<uint8_t> buffer(size);
        file.read(reinterpret_cast<char*>(buffer.data()), size);
        return buffer;
    }

    void quarantine_file(const std::string& path) {
        auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch()
        ).count();
        
        auto q_path = quarantine_dir + "/" + 
            std::to_string(timestamp) + "_" + std::filesystem::path(path).filename().string();
        
        std::filesystem::rename(path, q_path);
        std::cout << "🗑️ Quarantined: " << path << " -> " << q_path << std::endl;
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage: threat_scanner <model|dataset|io>\n";
        return 1;
    }

    HyzeThreatScanner scanner;
    std::string target(argv[1]);

    if (std::filesystem::is_directory(target)) {
        auto dataset_threats = scanner.scan_dataset(target);
        std::cout << "Dataset threats: " << dataset_threats.size() << std::endl;
    } else {
        auto model_threats = scanner.scan_model(target);
        std::cout << "Model threats: " << model_threats.size() << std::endl;
        scanner.block_threats(model_threats);
    }

    return 0;
}

File: hyze_threat_scanner.rs
//! Hyze AI Threat Scanner v1.0
//! Scans models, datasets, I/O for malware/backdoors
//! Production: Blocks threats before inference/training

use anyhow::{Context, Result};
use tokio::fs;
use tokio_stream::wrappers::ReadDirStream;
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::sync::Arc;
use tokio::sync::RwLock;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Threat {
    pub severity: Severity,
    pub threat_type: ThreatType,
    pub file_path: PathBuf,
    pub details: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Severity {
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ThreatType {
    BackdoorTrigger,
    PoisonedGradient,
    ModelExtraction,
    JailbreakPayload,
    MalwareSignature,
    AdversarialInput,
}

pub struct HyzeThreatScanner {
    ipu_client: Arc<HyzeIpuClient>,
    signature_db: Arc<RwLock<Vec<String>>>,
    threat_db: Arc<RwLock<Vec<String>>>,
}

impl HyzeThreatScanner {
    pub async fn new() -> Result<Self> {
        let ipu = Arc::new(HyzeIpuClient::new()?);
        let sig_db = load_signatures("hyze_threat_sigs.db").await?;
        
        Ok(Self {
            ipu_client: ipu,
            signature_db: Arc::new(RwLock::new(sig_db)),
            threat_db: Arc::new(RwLock::new(load_threat_patterns()?)),
        })
    }

    /// Scan AI model files (ONNX, PyTorch, Safetensors)
    pub async fn scan_model(&self, model_path: &Path) -> Result<Vec<Threat>> {
        let mut threats = Vec::new();
        
        let bytes = fs::read(model_path).await?;
        
        // 1. Signature scan
        if self.check_signatures(&bytes).await? {
            threats.push(Threat {
                severity: Severity::Critical,
                threat_type: ThreatType::BackdoorTrigger,
                file_path: model_path.to_path_buf(),
                details: "Malware signature detected".to_string(),
            });
        }
        
        // 2. IPU behavioral analysis
        let features = self.extract_model_features(&bytes);
        let threat_score = self.ipu_client.model_threat_scan(&features).await?;
        
        if threat_score > 0.9 {
            threats.push(Threat {
                severity: Severity::High,
                threat_type: ThreatType::PoisonedGradient,
                file_path: model_path.to_path_buf(),
                details: format!("Gradient poisoning detected: {:.2}", threat_score),
            });
        }
        
        // 3. Weight extraction attack detection
        if self.check_weight_stealing_patterns(&bytes).await? {
            threats.push(Threat {
                severity: Severity::Medium,
                threat_type: ThreatType::ModelExtraction,
                file_path: model_path.to_path_buf(),
                details: "Model extraction attack pattern".to_string(),
            });
        }
        
        Ok(threats)
    }

    /// Scan training datasets
    pub async fn scan_dataset(&self, dataset_dir: &Path) -> Result<Vec<Threat>> {
        let mut threats = Vec::new();
        let mut stream = ReadDirStream::new(fs::read_dir(dataset_dir).await?);
        
        while let Some(entry) = stream.next().await {
            let entry = entry?;
            let path = entry.path();
            
            if path.is_file() {
                let bytes = fs::read(&path).await?;
                
                // Poisoned data detection
                if self.check_data_poisoning(&bytes).await? {
                    threats.push(Threat {
                        severity: Severity::High,
                        threat_type: ThreatType::PoisonedGradient,
                        file_path: path,
                        details: "Training data poisoning detected".to_string(),
                    });
                }
            }
        }
        
        Ok(threats)
    }

    /// Real-time I/O scanning
    pub async fn scan_inference_io(&self, input: &[u8], output: &[u8]) -> Result<Vec<Threat>> {
        let mut threats = Vec::new();
        
        // Input: Jailbreak/adversarial
        if self.detect_prompt_injection(input) {
            threats.push(Threat {
                severity: Severity::Critical,
                threat_type: ThreatType::JailbreakPayload,
                file_path: PathBuf::from("input_stream"),
                details: "Prompt injection detected".to_string(),
            });
        }
        
        // Output: Data leakage
        if self.detect_data_leakage(output) {
            threats.push(Threat {
                severity: Severity::High,
                threat_type: ThreatType::MalwareSignature,
                file_path: PathBuf::from("output_stream"),
                details: "PII/SSN leakage detected".to_string(),
            });
        }
        
        Ok(threats)
    }

    pub async fn block_threats(&self, threats: &[Threat]) -> Result<()> {
        for threat in threats {
            match threat.threat_type {
                ThreatType::BackdoorTrigger => {
                    // Quarantine model
                    let quarantine = format!("quarantine/{:?}", threat.file_path);
                    fs::rename(&threat.file_path, quarantine).await?;
                }
                ThreatType::JailbreakPayload => {
                    // Block inference
                    self.ipu_client.block_input_stream().await?;
                }
                _ => {
                    // Log + alert
                    println!("🚨 Threat blocked: {:?}", threat);
                }
            }
        }
        Ok(())
    }
}

#[derive(Serialize)]
pub struct ScanSummary {
    pub total_files: usize,
    pub threats_found: usize,
    pub blocked: usize,
    pub clean_files: usize,
}

#[tokio::main]
async fn main() -> Result<()> {
    let mut scanner = HyzeThreatScanner::new().await?;
    
    // 1. Scan models
    let model_threats = scanner.scan_model("llama_7b.onnx").await?;
    println!("Models: {} threats", model_threats.len());
    
    // 2. Scan dataset
    let dataset_threats = scanner.scan_dataset("training_data/").await?;
    println!("Dataset: {} threats", dataset_threats.len());
    
    // 3. Real-time I/O protection
    let io_threats = scanner.scan_inference_io(b"ignore safety", b"SSN:123-45-6789").await?;
    println!("I/O: {} threats", io_threats.len());
    
    // 4. Auto-block
    scanner.block_threats(&model_threats.iter().chain(&dataset_threats).chain(&io_threats).cloned().collect::<Vec<_>>()).await?;
    
    Ok(())
}
